Firmware/communication/CircularBuffer.hpp

```diff
@@ -0,0 +1,333 @@
+*
+* Project       :
+*
+* Author        : S.Gilbert
+*
+* Created on    : 09 Dec 2019
+*
+*
+*******************************************************************************/
+/* Define to prevent recursive inclusion -------------------------------------*/
+#ifndef CIRCULAR_BUFFER_H
+#define CIRCULAR_BUFFER_H
+
+/*******************************************************************************
+INCLUDES
+*******************************************************************************/
+
+#include <string.h>
+#include <stddef.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+/*******************************************************************************
+NAMESPACE
+*******************************************************************************/
+
+/*******************************************************************************
+DEFINITIONS
+*******************************************************************************/
+
+typedef enum
+{
+	CB_OK = 0,
+	CB_FAIL,
+	CB_SIZE_ERROR,
+	CB_POINTER_ERROR
+}CB_Status_t;
+
+/*******************************************************************************
+TYPES
+*******************************************************************************/
+
+/*******************************************************************************
+GLOBAL VARIABLES
+*******************************************************************************/
+
+/*******************************************************************************
+FUNCTION PROTOTYPES
+*******************************************************************************/
+
+template <class dataType_t>
+class CCBBuffer
+{
+    typedef struct
+    {
+        size_t numElements;                                                     /* maximum number of elements */
+        int32_t start;                                                          /* index of oldest element */
+        int32_t end;                                                            /* index at which to write new element */
+    }CBTracker_t;
+
+    typedef struct
+    {
+        CBTracker_t tracker;                                                    /* buffer size and start/end counters */
+        dataType_t * pArray;                                                    /* pointer to array of elements. to be set on buffer init */
+    }circularBuffer_t;
+public:
+
+public:
+
+    CCBBuffer(dataType_t * pArray = nullptr, const size_t size = 0, const bool useMutex = false, const bool overwriteOldData = false);
+    ~CCBBuffer() = default;
+    bool isFull();
+    bool isEmpty();
+    size_t remainingSpace();
+    size_t remainingSpaceLinear();
+    size_t usedSpace();
+    size_t usedSpaceLinear();
+    bool flushBuffer();
+    int32_t write(const dataType_t * const elem, size_t length);
+    int32_t read(dataType_t * elem, size_t length);
+    bool readNewest(dataType_t * pData);
+
+private:
+    circularBuffer_t m_cb;
+    bool m_useMutex;
+    bool m_overwriteOldData;
```

Comment:
m_overwriteOldData is known for the current project and as it stands but templating it would mean it couldn't be changed during runtime. Also, templating it would cause duplication of code so it may become a speed/size optimisation issue.
m_cb.tracker.numElements was to be flexible to allow for changes in buffer size. Not needed here but has been used in a dynamic filter library. If however, you want this as a templated argument then the whole memory allocation could be internalised and the pointer and size allocation in the constructor could be removed.

Reply:
Technically it's possible to have one templated implementation to cover both static and dynamic use cases. [Here's a quick test](https://godbolt.org/z/rrAzMG). In the static case we see that the compiler is indeed aware that `size_` is a power of two, as it resolves `divide_by_size` to a simple left shift. With this approach the allocation would still have to be external which I think is quite ok.
I'm not too concerned about the size as long as we're not using 10 different instantiations throughout the firmware.
In general it certainly makes sense for you to use the same implementation throughout your projects but in any case it's somewhat likely that they will diverge sooner or later.